common/backoff/retry.go

```diff
@@ -89,23 +89,35 @@ func NewConcurrentRetrier(retryPolicy RetryPolicy) *ConcurrentRetrier {
 }
 
 // Retry function can be used to wrap any call with retry logic using the passed in policy
+// The returned error will be preferred to a previous one if one exists. That's because the
+// very last error is very likely a timeout error, and it's not useful for logging/troubleshooting
 func Retry(operation Operation, policy RetryPolicy, isRetryable IsRetryable) error {
+	var prevErr error
 	var err error
 	var next time.Duration
 
 	r := NewRetrier(policy, SystemClock)
 	for {
+		// record the previous error before an operation
+		prevErr = err
+
 		// operation completed successfully.  No need to retry.
 		if err = operation(); err == nil {
 			return nil
 		}
 
 		if next = r.NextBackOff(); next == done {
+			if prevErr != nil {
+				return prevErr
+			}
 			return err
 		}
 
 		// Check if the error is retryable
 		if isRetryable != nil && !isRetryable(err) {
+			if prevErr != nil {
+				return prevErr
```

Comment:
This could be confused to the caller if the err is non retryable but the returned prev error is retryable

Reply:
Retryable error could also be returned, if that's the last attempt. So it shouldn't be confusing.
We did the samething on the client side:https://github.com/uber-go/cadence-client/pull/878 
Because the last error usually is just timeout error because of maxing out the timeout value. It's useless.

Reply:
Essentially, operations with retries got a lot of errors returned, and we need to decide which one. Technically anyone should make sense. Returning last one is the simplest approach, but not useful for handling and logging.